transformer_engine/jax/cpp_extensions/gemm.py

```diff
@@ -0,0 +1,1340 @@
+    out_shape = out_shape_batched if batched_output else out_shape_2d
+
+    if out is None:
+        out = jnp.zeros(out_shape, dtype=lhs.dtype)
+
+    if extra_out is None:
+        extra_out_shape = 0
+        if comm_overlap_config is not None and comm_overlap_config["method"] != "bulk":
+            comm_type = comm_overlap_config["comm_type"]
+            if comm_type == tex.CommOverlapType.AG:
+                extra_out_shape = list(lhs.shape).copy()
+            elif comm_type == tex.CommOverlapType.RS:
+                extra_out_shape = list(out_shape).copy()
+        extra_out = jnp.zeros(extra_out_shape, dtype=lhs.dtype)
+
+    if not fuse_bias:
+        bias = jnp.zeros(0, dtype=lhs.dtype)
+    elif grad:
+        bias = jnp.zeros(out_shape[-1], dtype=lhs.dtype)
+    else:
+        assert bias is not None, "Missing bias in forward GEMM when bias epilogue is enabled."
+
+    if not fuse_gelu:
+        gelu_input = jnp.zeros(0, dtype=lhs.dtype)
+    elif grad:
+        assert (
+            gelu_input is not None
+        ), "Backward GEMM with dGELU epilogue requires pre-GELU output from forward GEMM."
+    elif gelu_input is None:
+        gelu_input = jnp.zeros(out_shape_2d, dtype=lhs.dtype)
+
+    (
+        out,
+        _,  # out_amax in FP8 GEMM
+        _,  # out_scale in FP8 GEMM
+        pre_gelu_out,
+        bias_grad,
+        extra_out,
+    ) = CollectiveGemmPrimitive.outer_primitive.bind(
+        lhs,
+        dummy_fp8_meta,
+        rhs,
+        dummy_fp8_meta,
+        bias,
+        gelu_input,
+        out,
+        dummy_fp8_meta,
+        dummy_fp8_meta,
+        extra_out,
+        batched_output=batched_output,
+        contracting_dims=contracting_dims,
+        fuse_gelu=fuse_gelu,
+        fuse_bias=fuse_bias,
+        grad=grad,
+        accumulate=accumulate,
+        use_split_accumulator=use_split_accumulator,
+        comm_overlap_config=comm_overlap_config,
+        sharded_abstract=False,
+    )
+
+    if grad:
+        return out, pre_gelu_out, bias_grad, extra_out
+    else:
+        return out, pre_gelu_out, extra_out
+
+
+def fp8_gemm_impl(
+    lhs: ArrayLike,
+    lhs_scale_inv: ArrayLike,
+    rhs_t: ArrayLike,
+    rhs_scale_inv: ArrayLike,
+    bias: Optional[ArrayLike] = None,
+    gelu_input: Optional[ArrayLike] = None,
+    out: Optional[ArrayLike] = None,
+    extra_out: Optional[ArrayLike] = None,
+    out_amax: Optional[ArrayLike] = None,
+    out_scale: Optional[ArrayLike] = None,
+    out_dtype: jnp.dtype = jnp.bfloat16,
+    batched_output: bool = False,
+    fuse_gelu: bool = False,
+    fuse_bias: bool = False,
+    accumulate: bool = False,
+    use_split_accumulator: bool = False,
+    comm_overlap_config: Optional[dict] = None,
+) -> Tuple[ArrayLike, ...]:
+    """FP8 mat-mul with `nvte_cublas_gemm()` custom op."""
+    out_shape_batched = (*lhs.shape[:-2], lhs.shape[-1], rhs_t.shape[-1])
```

Comment:
The output shape looks wrong to me. Should be (*lhs.shape[:-2], lhs.shape[-2], rhs_t.shape[-2]).
@denera